### PR TITLE
Add location (filename and line number) to errors

### LINKED ISSUE #85: Add the line to the error
As a user pointed out in our gitter, it is possible to write some information together with an error about where the error happens. Typesafe config has [`ConfigValue::origin`](http://typesafehub.github.io/config/latest/api/com/typesafe/config/Config.html#origin--) which is supposed to be used exactly for this.

### PR DESCRIPTION
**Pull request by jcazevedo**
This PR adds the location (filename and line number) to `ConfigReaderFailure`. The location is an `Option` since it might not always be available (we don't always load configurations from a file and some errors can't be pinned to a specific location).

This PR attempts to fix #85.

This is still WIP since I'd like to add more tests and improve the documentation.

### CODE REVIEW

**Comment by jcazevedo** [comment]
I've improved the documentation of the sealed family of `ConfigReaderFailure`s and added a description to the README on how to provide locations to failures.

@melrief and @leifwickland: I think this is ready to be reviewed.

**Comment by melrief** [comment]
I will review it tomorrow, thanks for the work @jcazevedo !

File: core/docs/error-handling.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+## Error handling
+
+PureConfig provides some semantically rich case classes to describe failures
```
Comment (leifwickland) [inline core/docs/error-handling.md]: Suggestion: I'd reword this sentence as:

> The `pureconfig.error` package provides semantically rich case classes to describe failures reading from a `ConfigValue`.

Feel free to prefer your wording.

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala [hunk 2]
```diff
@@ -5,14 +5,152 @@ package pureconfig.error
...
-final case class WrongTypeForKey(typ: String, expectedTyp: String, key: String) extends ConfigReaderFailure
-final case class ThrowableFailure(throwable: Throwable) extends ConfigReaderFailure
-final case class EmptyStringFound(typ: String) extends ConfigReaderFailure
-final case class NoValidCoproductChoiceFound(value: ConfigValue) extends ConfigReaderFailure
\ No newline at end of file
+/**
+ * The physical location of a ConfigValue, represented by a file name and a line
+ * number
+ *
+ * @param filename the complete pathname to the file that contains a given
```
Comment (leifwickland) [inline core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala:12]: If the config file is inside of a JAR does this become some useful value like:
`path/to/the.jar#resource/path.conf`?

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala [hunk 3]
```diff
@@ -5,14 +5,152 @@ package pureconfig.error
...
+ *                 ConfigValue
+ * @param lineNumber the line number (starting at 0), where the given
+ *                   ConfigValue definition starts
+ */
+case class ConfigValueLocation(filename: String, lineNumber: Int) {
+  def description: String = s"(file:$filename: $lineNumber)"
```
Comment (leifwickland) [inline core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala]: Just a suggestion: 
I think I'd rather this be 
```
s"($filename:$lineNumber)"
```
which seems like a more traditional format for an error report.

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala [hunk 4]
```diff
@@ -5,14 +5,152 @@ package pureconfig.error
...
+   *         possible to derive a location. It is not possible to derive
+   *         locations from ConfigValues that are not in files or for
+   *         ConfigValues that are null.
+   */
+  def apply(cv: ConfigValue): Option[ConfigValueLocation] =
+    Option(cv).flatMap { v =>
+      val origin = v.origin()
+      if (origin.filename != null && origin.lineNumber != -1)
+        // I believe lineNumber starts at 1 in ConfigOrigin
+        Some(ConfigValueLocation(origin.filename, origin.lineNumber - 1))
```
Comment (leifwickland) [inline core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala]: If ConfigOrigin's line numbering starts at 1 and you're subtracting 1, does that mean you're trying to get zero-based line numbering?
If so I object. I would prefer one-based line numbering.

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala [hunk 5]
```diff
@@ -5,14 +5,152 @@ package pureconfig.error
...
+ */
+sealed abstract class ConfigReaderFailure {
+  /**
+   * The optional location of the ConfigReaderFailure
+   */
+  def location: Option[ConfigValueLocation]
+}
+
+/**
+ * A failure representing the inability of converting a null value. Since a null
```
Comment (leifwickland) [inline core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala]: Nitpick: Here and below in a place or two, I'd suggest the phrase "the inability to convert" instead of "the inability of converting".

File: core/src/test/scala/pureconfig/ConfigReaderFailureLocationSuite.scala [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+      """java.lang.NumberFormatException: For input string: "hello"""",
+      Some(ConfigValueLocation(workingDir + file, 8))))
+  }
+
+  it should "show proper error location when loading from multiple files" in {
+    import pureconfig.syntax._
+    case class Conf(a: Int, b: String, c: Int)
+
+    val workingDir = getClass.getResource("/").getFile
+    val file1 = "conf/error2/a.conf"
```
Comment (leifwickland) [inline core/src/test/scala/pureconfig/ConfigReaderFailureLocationSuite.scala]: Suggestion: Rename the `error2` and `error1` parts of the paths these test files are in to something like `errorLocationReporting` to try to tie them to their purpose.

File: core/src/test/scala/pureconfig/ConfigReaderFailureLocationSuite.scala [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+    val file2 = "conf/error2/b.conf"
+    val conf = ConfigFactory.load(file1).withFallback(ConfigFactory.load(file2)).root()
+
+    val failures = conf.get("conf").to[Conf].left.value.toList
+    failures should have size 1
+    failures should contain(CannotConvert(
+      "string",
+      "Int",
+      """java.lang.NumberFormatException: For input string: "string"""",
+      Some(ConfigValueLocation(workingDir + file2, 1))))
```
Comment (leifwickland) [inline core/src/test/scala/pureconfig/ConfigReaderFailureLocationSuite.scala]: That seems like it should be line 2, not 1, to me.

**Review by leifwickland — CHANGES_REQUESTED** [review]
I left a few suggestions.
I'd sure line numbering for error reporting to be one-based.

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala (hunk 4 above)
Comment (jcazevedo) [reply core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala]: I agree. I'm not sure what I was thinking. 1-based makes much more sense. I'll fix it later today.

File: core/docs/error-handling.md (hunk 1 above)
Comment (jcazevedo) [reply core/docs/error-handling.md]: I've reworded it as per your suggestion in 402a1d8.

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala (hunk 2 above)
Comment (jcazevedo) [reply core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala:12]: Yes. It will be something like `path/to/the.jar!/path.conf`.

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala (hunk 3 above)
Comment (jcazevedo) [reply core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala]: I've modified it in e1f4694. I don't have a strong preference either way, so I'm OK to go with the more traditional format.

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala (hunk 4 above)
Comment (jcazevedo) [reply core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala]: Fixed it in a8248b7.

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala (hunk 5 above)
Comment (jcazevedo) [reply core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala]: Modified the wording as per your suggestion in 402a1d8.

File: core/src/test/scala/pureconfig/ConfigReaderFailureLocationSuite.scala (hunk 6 above)
Comment (jcazevedo) [reply core/src/test/scala/pureconfig/ConfigReaderFailureLocationSuite.scala]: I've scoped the test files used by this suite under the `configFailureLocation` folder in 3c08eb3.

File: core/src/test/scala/pureconfig/ConfigReaderFailureLocationSuite.scala (hunk 7 above)
Comment (jcazevedo) [reply core/src/test/scala/pureconfig/ConfigReaderFailureLocationSuite.scala]: Yes, you're right. This had to do with the line numbering not being one-based. I've fixed it in a8248b7.

**Comment by jcazevedo** [comment]
@leifwickland: Thanks for your review! I believe I've addressed all your comments.

**Review by leifwickland — APPROVED** [review]
Awesome work, @jcazevedo! This is a terrific feature.
Thanks for considering my suggestions.

File: core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala [hunk 8]
```diff
@@ -5,14 +5,151 @@ package pureconfig.error
...
+        None
+    }
+}
+
+/**
+ * A representation of a failure that might be raised from reading a
+ * ConfigValue. The failure contains an optional location of the ConfigValue
+ * that raised the error.
+ */
+sealed abstract class ConfigReaderFailure {
```
Comment (melrief) [inline core/src/main/scala/pureconfig/error/ConfigReaderFailure.scala:49]: @jcazevedo thanks a lot for adding documentation to the failures!

**Review by melrief — APPROVED** [review]
Amazing job @jcazevedo !
